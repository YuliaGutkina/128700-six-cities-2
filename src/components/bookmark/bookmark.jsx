import React, {PureComponent} from 'react';
import PropTypes from "prop-types";
import classNames from "classnames";
import {withRouter} from "react-router-dom";
import {connect} from "react-redux";

import {receiveUserDataSelector} from "../../reducer/user/selectors";
import {Operation} from "../../reducer/data/data";
import {getFavoriteStatusSelector} from "../../reducer/data/selectors";


class Bookmark extends PureComponent {
  constructor(props) {
    super(props);

    this._handleFavoriteClick = this._handleFavoriteClick.bind(this);
  }

  render() {
    const {offer, className, iconClassName, isLarge} = this.props;

    return <button
      className={classNames(
          `button`,
          className,
          offer.isFavorite ? `${className}--active` : null
      )}
      type="button"
      onClick={this._handleFavoriteClick}
    >
      <svg
        className={iconClassName}
        width={isLarge ? 31 : 18}
        height={isLarge ? 33 : 19}
      >
        <use xlinkHref="#icon-bookmark" />
      </svg>
      <span className="visually-hidden">To bookmarks</span>
    </button>;
  }

  _handleFavoriteClick() {
    const {offer, userData, history, onFavoriteClick} = this.props;

    if (!userData) {
      history.push(`/login`);
      return;
    }

    onFavoriteClick(offer);
  }
}

Bookmark.propTypes = {
  offer: PropTypes.shape({
    id: PropTypes.number,
    city: PropTypes.shape({
      name: PropTypes.string,
      location: PropTypes.shape({
        latitude: PropTypes.number,
        longitude: PropTypes.number,
        zoom: PropTypes.number
      })
    }),
    preview: PropTypes.string,
    images: PropTypes.arrayOf(PropTypes.string),
    title: PropTypes.string,
    isFavorite: PropTypes.bool,
    isPremium: PropTypes.bool,
    rating: PropTypes.number,
    type: PropTypes.string,
    bedrooms: PropTypes.number,
    maxAdults: PropTypes.number,
    price: PropTypes.number,
    goods: PropTypes.arrayOf(PropTypes.string),
    host: PropTypes.shape({
      id: PropTypes.number,
      isPro: PropTypes.bool,
      name: PropTypes.string,
      avatar: PropTypes.string
    }),
    description: PropTypes.string,
    location: PropTypes.shape({
      latitude: PropTypes.number,
      longitude: PropTypes.number,
      zoom: PropTypes.number
    })
  }).isRequired,
  onFavoriteClick: PropTypes.func.isRequired,
  userData: PropTypes.shape({
    id: PropTypes.number,
    email: PropTypes.string,
    name: PropTypes.string,
    avatar: PropTypes.string,
    isPro: PropTypes.bool
  }),
  history: PropTypes.object,
  isFavorite: PropTypes.bool,
  className: PropTypes.string,
  iconClassName: PropTypes.string,
  isLarge: PropTypes.bool
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  userData: receiveUserDataSelector(state),
  isFavorite: getFavoriteStatusSelector(state, ownProps.offer.id)
});

const mapDispatchToProps = (dispatch) => ({
  onFavoriteClick: (offer) => {
    dispatch(Operation.toggleFavoriteStatus(offer));
  },
});

export {Bookmark};
export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Bookmark));
