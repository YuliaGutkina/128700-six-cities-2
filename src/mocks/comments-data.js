export const commentsData = [
  {
    id: 1,
    user: {
      id: 4,
      isPro: false,
      name: `Max`,
      avatar: `img/1.png`
    },
    rating: 4,
    comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    date: `2019-05-08T14:13:56.569Z`
  },
  {
    id: 2,
    user: {
      id: 4,
      isPro: false,
      name: `Max`,
      avatar: `img/1.png`
    },
    rating: 4,
    comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    date: `2019-05-08T14:13:56.569Z`
  },
  {
    id: 3,
    user: {
      id: 4,
      isPro: false,
      name: `Max`,
      avatar: `img/1.png`
    },
    rating: 4,
    comment: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    date: `2019-05-08T14:13:56.569Z`
  }
];
