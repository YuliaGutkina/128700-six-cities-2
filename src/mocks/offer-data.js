export const offersData = [
  {
    id: 1,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    preview: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatar: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8
    }
  },
  {
    id: 2,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    preview: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatar: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8
    }
  },
  {
    id: 3,
    city: {
      name: `Tel Aviv`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    preview: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatar: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8
    }
  },
  {
    id: 4,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    preview: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatar: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8
    }
  },
  {
    id: 5,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    preview: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatar: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8
    }
  },
  {
    id: 6,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    preview: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatar: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8
    }
  },
  {
    id: 7,
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    preview: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatar: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8
    }
  },
  {
    id: 8,
    city: {
      name: `Paris`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    preview: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatar: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8
    }
  },
  {
    id: 9,
    city: {
      name: `Paris`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    preview: `img/1.png`,
    images: [`img/1.png`, `img/2.png`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatar: `img/1.png`
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.35514938496378,
      longitude: 4.673877537499948,
      zoom: 8
    }
  }
];
