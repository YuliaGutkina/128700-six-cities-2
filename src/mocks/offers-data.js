export const offersData = [
  {
    city: `Amsterdam`,
    initialCoordinates: [52.38333, 4.9],
    places: [
      {
        title: `Beautiful & luxurious apartment at great location`,
        type: `Apartment`,
        isPremium: true,
        src: `img/apartment-01.jpg`,
        price: 120,
        rating: 93,
        coordinates: [52.3909553943508, 4.85309666406198]
      },
      {
        title: `Wood and stone place`,
        type: `Private room`,
        isPremium: false,
        src: `img/room.jpg`,
        price: 80,
        rating: 80,
        coordinates: [52.369553943508, 4.85309666406198]
      },
      {
        title: `Canal View Prinsengracht`,
        type: `Apartment`,
        isPremium: false,
        src: `img/apartment-02.jpg`,
        price: 132,
        rating: 80,
        coordinates: [52.3909553943508, 4.929309666406198]
      },
      {
        title: `Nice, cozy, warm big bed apartment`,
        type: `Apartment`,
        isPremium: true,
        src: `img/apartment-03.jpg`,
        price: 180,
        rating: 100,
        coordinates: [52.3809553943508, 4.939309666406198]
      }
    ]
  },
  {
    city: `Paris`,
    initialCoordinates: [48.864716, 2.349014],
    places: []
  },
  {
    city: `Hamburg`,
    initialCoordinates: [53.551086, 9.993682],
    places: []
  }
];
