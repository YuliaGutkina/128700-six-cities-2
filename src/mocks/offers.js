export const offers = [
  {
    title: `Beautiful & luxurious apartment at great location`,
    type: `Apartment`,
    isPremium: true,
    src: `img/apartment-01.jpg`,
    price: 120,
    rating: 93
  },
  {
    title: `Wood and stone place`,
    type: `Private room`,
    isPremium: false,
    src: `img/room.jpg`,
    price: 80,
    rating: 80
  },
  {
    title: `Canal View Prinsengracht`,
    type: `Apartment`,
    isPremium: false,
    src: `img/apartment-02.jpg`,
    price: 132,
    rating: 80
  },
  {
    title: `Nice, cozy, warm big bed apartment`,
    type: `Apartment`,
    isPremium: true,
    src: `img/apartment-03.jpg`,
    price: 180,
    rating: 100
  }
];
