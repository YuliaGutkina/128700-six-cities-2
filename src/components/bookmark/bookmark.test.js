import React from "react";
import ShallowRenderer from 'react-test-renderer/shallow';

import {offersData} from "../../mocks/offer-data";
import {Bookmark} from "./bookmark";


it(`Bookmark correctly renders after relaunch`, () => {
  const renderer = new ShallowRenderer();
  renderer.render(
      <Bookmark
        offer={offersData[0]}
        onFavoriteClick={jest.fn()}
        userData={{fake: true}}
        history={{push: jest.fn}}
        isFavorite={true}
      />);
  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});
