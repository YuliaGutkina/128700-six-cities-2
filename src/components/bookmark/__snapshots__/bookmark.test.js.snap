// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark correctly renders after relaunch 1`] = `
<button
  className="button"
  onClick={[Function]}
  type="button"
>
  <svg
    height={19}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;
